File: src/article_scraper_lab/security/url_policy.py
"""Strict URL parsing, public-IP validation, and connection targets."""

from __future__ import annotations

import ipaddress
import socket
from collections.abc import Callable, Iterable
from dataclasses import dataclass
from urllib.parse import urlsplit, urlunsplit

from article_scraper_lab.errors import UnsafeUrlError

Resolver = Callable[[str, int], Iterable[str]]


def system_resolver(hostname: str, port: int) -> set[str]:
    try:
        records = socket.getaddrinfo(hostname, port, type=socket.SOCK_STREAM)
    except socket.gaierror:
        raise UnsafeUrlError("Hostname tidak dapat di-resolve") from None
    addresses = {record[4][0].split("%", 1)[0] for record in records}
    if not addresses:
        raise UnsafeUrlError("Hostname tidak memiliki alamat IP")
    return addresses


@dataclass(frozen=True, slots=True)
class UrlTarget:
    url: str
    scheme: str
    hostname: str
    port: int
    host_header: str
    request_target: str
    resolved_ips: tuple[str, ...]


class UrlPolicy:
    def __init__(
        self,
        *,
        allow_http: bool = False,
        allowed_domains: tuple[str, ...] = (),
        resolver: Resolver = system_resolver,
    ) -> None:
        self._allow_http = allow_http
        self._allowed_domains = tuple(domain.lower().rstrip(".") for domain in allowed_domains)
        self._resolver = resolver

    def validate(self, raw_url: str) -> UrlTarget:
        if not raw_url or len(raw_url) > 2048:
            raise UnsafeUrlError("URL kosong atau terlalu panjang")
        if any(ord(character) < 32 for character in raw_url):
            raise UnsafeUrlError("URL mengandung karakter kontrol")
        try:
            parsed = urlsplit(raw_url.strip())
            port = parsed.port
        except ValueError:
            raise UnsafeUrlError("Format URL tidak valid") from None
        scheme = parsed.scheme.lower()
        if scheme not in {"http", "https"}:
            raise UnsafeUrlError("Hanya URL HTTP/HTTPS yang didukung")
        if scheme == "http" and not self._allow_http:
            raise UnsafeUrlError("HTTP tidak diizinkan; gunakan HTTPS")
        if parsed.username is not None or parsed.password is not None:
            raise UnsafeUrlError("Credential tidak boleh ditanam di dalam URL")
        if not parsed.hostname:
            raise UnsafeUrlError("URL tidak memiliki hostname")

        try:
            hostname = parsed.hostname.encode("idna").decode("ascii").lower().rstrip(".")
        except UnicodeError:
            raise UnsafeUrlError("Hostname tidak valid") from None
        self._validate_domain_allowlist(hostname)

        expected_port = 443 if scheme == "https" else 80
        port = port or expected_port
        if port != expected_port:
            raise UnsafeUrlError("Hanya port standar 443/80 yang diizinkan")

        addresses = self._resolve(hostname, port)
        for address in addresses:
            self._validate_public_ip(address)

        is_ipv6 = ":" in hostname
        display_host = f"[{hostname}]" if is_ipv6 else hostname
        host_header = display_host if port == expected_port else f"{display_host}:{port}"
        path = parsed.path or "/"
        request_target = f"{path}?{parsed.query}" if parsed.query else path
        normalized_url = urlunsplit((scheme, host_header, path, parsed.query, ""))
        return UrlTarget(
            url=normalized_url,
            scheme=scheme,
            hostname=hostname,
            port=port,
            host_header=host_header,
            request_target=request_target,
            resolved_ips=tuple(sorted(addresses)),
        )

    def _resolve(self, hostname: str, port: int) -> set[str]:
        try:
            literal = ipaddress.ip_address(hostname)
        except ValueError:
            addresses = set(self._resolver(hostname, port))
            if not addresses:
                raise UnsafeUrlError("Hostname tidak memiliki alamat IP")
            return addresses
        return {str(literal)}

    @staticmethod
    def _validate_public_ip(address: str) -> None:
        try:
            parsed = ipaddress.ip_address(address.split("%", 1)[0])
        except ValueError:
            raise UnsafeUrlError("DNS menghasilkan alamat IP tidak valid") from None
        if not parsed.is_global:
            raise UnsafeUrlError("URL mengarah ke jaringan lokal, privat, atau khusus")

    def _validate_domain_allowlist(self, hostname: str) -> None:
        if not self._allowed_domains:
            return
        if not any(
            hostname == domain or hostname.endswith(f".{domain}")
            for domain in self._allowed_domains
        ):
            raise UnsafeUrlError("Domain tidak terdaftar pada ALLOWED_DOMAINS")

File: internal/security/urlpolicy.go
package security

import (
	"context"
	"fmt"
	"net"
	"net/url"
	"sort"
	"strings"

	"golang.org/x/net/idna"
)

type Resolver interface {
	LookupIP(context.Context, string, string) ([]net.IP, error)
}

type Target struct {
	URL      *url.URL
	Hostname string
	Port     string
	IPs      []net.IP
}

type Policy struct {
	AllowHTTP      bool
	AllowedDomains []string
	Resolver       Resolver
}

func (p Policy) Validate(ctx context.Context, raw string) (Target, error) {
	if raw == "" || len(raw) > 2048 {
		return Target{}, fmt.Errorf("URL kosong atau terlalu panjang")
	}
	for _, r := range raw {
		if r < 32 {
			return Target{}, fmt.Errorf("URL mengandung karakter kontrol")
		}
	}
	u, err := url.Parse(strings.TrimSpace(raw))
	if err != nil || u.Hostname() == "" {
		return Target{}, fmt.Errorf("format URL tidak valid")
	}
	scheme := strings.ToLower(u.Scheme)
	if scheme != "http" && scheme != "https" {
		return Target{}, fmt.Errorf("hanya URL HTTP/HTTPS yang didukung")
	}
	if scheme == "http" && !p.AllowHTTP {
		return Target{}, fmt.Errorf("HTTP tidak diizinkan; gunakan HTTPS")
	}
	if u.User != nil {
		return Target{}, fmt.Errorf("credential tidak boleh ditanam di dalam URL")
	}
	host, err := idna.Lookup.ToASCII(strings.TrimSuffix(u.Hostname(), "."))
	if err != nil {
		return Target{}, fmt.Errorf("hostname tidak valid")
	}
	host = strings.ToLower(host)
	if !p.allowed(host) {
		return Target{}, fmt.Errorf("domain tidak terdaftar pada ALLOWED_DOMAINS")
	}
	expected := "443"
	if scheme == "http" {
		expected = "80"
	}
	port := u.Port()
	if port == "" {
		port = expected
	}
	if port != expected {
		return Target{}, fmt.Errorf("hanya port standar 443/80 yang diizinkan")
	}
	resolver := p.Resolver
	if resolver == nil {
		resolver = net.DefaultResolver
	}
	ips, err := resolver.LookupIP(ctx, "ip", host)
	if err != nil || len(ips) == 0 {
		return Target{}, fmt.Errorf("hostname tidak dapat di-resolve")
	}
	for _, ip := range ips {
		if !isPublic(ip) {
			return Target{}, fmt.Errorf("URL mengarah ke jaringan lokal, privat, atau khusus")
		}
	}
	sort.Slice(ips, func(i, j int) bool { return ips[i].String() < ips[j].String() })
	u.Scheme = scheme
	u.Host = host
	if strings.Contains(host, ":") {
		u.Host = "[" + host + "]"
	}
	if u.Path == "" {
		u.Path = "/"
	}
	u.Fragment = ""
	return Target{URL: u, Hostname: host, Port: port, IPs: ips}, nil
}

var specialNetworks = []*net.IPNet{
	mustCIDR("0.0.0.0/8"), mustCIDR("100.64.0.0/10"), mustCIDR("192.0.0.0/24"),
	mustCIDR("192.0.2.0/24"), mustCIDR("198.18.0.0/15"), mustCIDR("198.51.100.0/24"),
	mustCIDR("203.0.113.0/24"), mustCIDR("224.0.0.0/4"), mustCIDR("240.0.0.0/4"),
	mustCIDR("2001:db8::/32"), mustCIDR("2001:10::/28"), mustCIDR("ff00::/8"),
}

func isPublic(ip net.IP) bool {
	if !ip.IsGlobalUnicast() || ip.IsPrivate() || ip.IsLoopback() || ip.IsLinkLocalUnicast() || ip.IsUnspecified() {
		return false
	}
	for _, network := range specialNetworks {
		if network.Contains(ip) {
			return false
		}
	}
	return true
}

func mustCIDR(raw string) *net.IPNet {
	_, network, err := net.ParseCIDR(raw)
	if err != nil {
		panic(err)
	}
	return network
}

func (p Policy) allowed(host string) bool {
	if len(p.AllowedDomains) == 0 {
		return true
	}
	for _, d := range p.AllowedDomains {
		d = strings.ToLower(strings.TrimSuffix(d, "."))
		if host == d || strings.HasSuffix(host, "."+d) {
			return true
		}
	}
	return false
}
